Rework README file (#14)
wojpawlik/deno2node: README.md

# deno2node

Compile your [Deno] project to run on [Node.js].

## CLI Usage from Deno

No installation needed. Simply `cd` into the directory of your project, and run:

```sh

> deno run --unstable --allow-read=. --allow-write=. \

Author: +1 but please remember to adjust this example after https://github.com/denoland/deno/issues/11340

    https://deno.land/x/deno2node/src/cli.ts <tsConfigFilePath>
```

You need to substitute `<tsConfigFilePath>` by a path to your `tsconfig.json`
file. `deno2node` passes it on to the shipped `tsc` for compiling your code.

## CLI Usage from Node.js

As a by-product of end-to-end testing, a Node.js build is also available:

```sh
npm install --save-dev --save-prefix='~' deno2node
```

Now add a script to `package.json` so you can run it with `npm run build`:

```jsonc
{
  // yada yada ...
  "scripts": {

>     "build": "deno2node <tsConfigFilePath>"

Author: `prepare` has several use cases, as it is aulso invoked automatically. Users who know this are able to adjust the example, others will appreciate a descriptive name.

That being ssid, it's obviously your choice :wink:

  }
}
```

Instead of adding a script, you can also run it directly:

```sh
npx deno2node <tsConfigFilePath>
```

## Motivation

Writing code for Deno is fun. It's so much simpler.

If you have tried it, you know what we mean.

If not, here is how you develop for Deno: Don't wait for TypeScript to
transpile. Just run it. Don't wait for unit tests. They are truly fast. Linting?
It is always instant. Zero configuration files. The tooling is standardised,
built-in, and works well together. No `package.json` to tediously fill out. No
more “did you try `rm -rf node_modules`?” Everything just works out of the box.
What a relief.

Try it, and you will know what we mean.

## How it Works

There are three main steps to this.

1. Transform the code base. This means:
   - Rewrite all import statements, e.g. to handle explicit vs. implicit file
     extensions
   - Reverse npm package transformations from `skypack.dev`, `esm.sh`,

>      `deno.land/std`, and `nest.land/std`.

Reviewer: ```suggestion
     `deno.land/std/node`, and `nest.land/std`.
```

I can't find a way to link to _lastest_ `next.land/std/node` 😩

Author: Oops, I did not read the source closely enough. Go ahead

inject them for you:

```jsonc
{
  "deno2node": {
    "shim": "src/shim.node.ts" // path to shim file, relative to tsconfig
  }
}
```

If you simply want to shim all Deno globals, you can install the `deno.ns`
package. Then do:

```ts
// src/shim.node.ts
export * from "deno.ns";
```

That package was specifically created for this use case.

### Runtime-specific code

In same cases you may want to have two different implementations, depending on
whether you're running on Deno or on Node. When shimming is not enough, you can
provide a Node-specific `<anything>.node.ts` and a Deno-specific
`<anything>.deno.ts` version of any file. They need to reside next to each other
in the same directory.

`deno2node` will ignore the Deno version and rewrite imports to use the Node.js
version instead. Thus, the Deno-specific file will not be part of the build
output.

For example, provide `greet.deno.ts` for Deno:

```ts
// src/greet.deno.ts
export function greet() {
  console.log("Hello Deno!");
  // access Deno-specific APIs here
}
```

Now, provide `greet.node.ts` for Node:

```ts
// src/greet.node.ts
export function greet() {
  console.log("Hello Node!");
  // access Node-specific APIs here
}

> ```

Reviewer: This needs a better example. Like hashing!
https://github.com/fromdeno/cache/blob/8a9fcd45ec8d473734c592d26af65432034332bf/src/deps.deno.ts#L3-L5
https://github.com/fromdeno/cache/blob/8a9fcd45ec8d473734c592d26af65432034332bf/src/deps.node.ts#L3-L6

Author: The point is just that there can be platform-specific code, and this example is perfectly obvious. Despite `console.log` being identical.

Finally, use it in `foo.ts`:

```ts
import { greet } from "./platform.deno.ts";

// Prints "Hello Deno!" on Deno,
// and "Hello Node!" on Node:
greet();
```

This technique has many uses. `deno2node` itself uses it to import from
https://deno.land/x. A plugin of the Telegram bot framework [`grammY`] uses it
to abstract away platform-specific APIs.

### Vendoring

If you import a module which has no npm equivalent, `deno2node` can extract the
code out of Deno's module cache. On Deno, this directory can be used for
vendoring (by specifying the `$DENO_DIR` environment variable), so `deno2node`
calls it `vendorDir`.

```jsonc
// tsconfig.json
{
  "deno2node": {
    "vendorDir": "src/.deno2node/vendor/" // path within `rootDir`
  }
}
```

> > Note that vendoring requires `--allow-env` in order to locate Deno cache.

Reviewer: It also needs read access to Deno cache, which the example code doesn't provide. But I think I'll simply add a permission prompt.

## CLI Usage from Node.js

As a by-product of end-to-end testing, a Node.js build is also available:

```sh
npm install --save-dev --save-prefix='~' deno2node
```

Now add a script to `package.json` so you can run it with `npm run build`:

```jsonc
{
  // yada yada ...
  "scripts": {
    "build": "deno2node <tsConfigFilePath>"
  }
}
```

Instead of adding a script, you can also run it directly:

```sh
npx deno2node <tsConfigFilePath>
```

## Motivation

Writing code for Deno is fun. It's so much simpler.

If you have tried it, you know what we mean.

If not, here is how you develop for Deno: Don't wait for TypeScript to
transpile. Just run it. Don't wait for unit tests. They are truly fast. Linting?
It is always instant. Zero configuration files. The tooling is standardised,
built-in, and works well together. No `package.json` to tediously fill out. No
more “did you try `rm -rf node_modules`?” Everything just works out of the box.
What a relief.

Try it, and you will know what we mean.

## How it Works

There are three main steps to this.

1. Transform the code base. This means:
   - Rewrite all import statements, e.g. to handle explicit vs. implicit file

>      extensions
>    - Reverse npm package transformations from `skypack.dev`, `esm.sh`,
>      `deno.land/std`, and `nest.land/std`.

Reviewer: ```suggestion
     extensions.
   - Use bare specifiers in `std/node`, https://esm.sh and https://skypack.dev imports.
```

Author: You can modify this PR as you wish, or not. Apply the changes you'd like to see—I do not wish to start discussions about this :)

Reviewer: Ok, although I'd like to ensure that I don't mess up your flair :)

I'll commit this, tweak a few more things, wait for your review, and then finally merge.

Author: I'm glad you think it has flair :D
Don't worry.

I can do the review once it's fine :)

# deno2node

Compile your [Deno] project to run on [Node.js].

> ![Because Deno's tooling is simpler than Node's](https://pbs.twimg.com/media/FBba11IXMAQB7pX?format=jpg)

Reviewer: ```suggestion
![Because Deno's tooling is way simpler than Node's](https://pbs.twimg.com/media/FBba11IXMAQB7pX?format=jpg)
```

> ## CLI Usage from Deno

Reviewer: ```suggestion
## CLI usage from Deno
```

Author: ```suggestion
## CLI Usage From Deno
```

No installation needed. Simply `cd` into the directory of your project, and run:

```sh
deno run --no-check --allow-read=. --allow-write=. \
  https://deno.land/x/deno2node/src/cli.ts <tsConfigFilePath>
```

You need to substitute `<tsConfigFilePath>` by a path to your `tsconfig.json`
file. `deno2node` passes it on to the shipped `tsc` for compiling your code.

> ## CLI Usage from Node.js

Reviewer: ```suggestion
## CLI usage from Node.js
```

Author: ```suggestion
## CLI Usage From Node.js
```

As a by-product of end-to-end testing, a Node.js build is also available:

```sh
# New fearues or TypeScript upgrades
# may alter output or diagnostics across minor versions.
npm install --save-dev --save-prefix='~' deno2node
```

Now add a script to `package.json` so you can run it with `npm run prepare`:

```jsonc
// @filename: package.json
{
  // yada yada ...
  "scripts": {
    "prepare": "deno2node <tsConfigFilePath>"
  }
}
```

You can also run it directly:

```sh
npx deno2node <tsConfigFilePath>
```

> ## How it works

Reviewer: ```suggestion
## How it Works
```

?
